File: core/plugin/__DEV/C++/source/n_gfUtilEulerScalarMath.cpp
#include "n_gfUtilEulerScalarMath.h"


// Configure a input attribute.
#define INPUT_ATTR(FNATTR)		\
	FNATTR.setWritable(true);	\
	FNATTR.setReadable(true);	\
	FNATTR.setStorable(true);	\
	FNATTR.setKeyable(true);	\

// Configure a output attribute.
#define OUTPUT_ATTR(FNATTR)		\
	FNATTR.setWritable(false);	\
	FNATTR.setReadable(true);	\
	FNATTR.setStorable(false);	\
	FNATTR.setKeyable(false);	\


// Constructor.
EulerScalarMath::EulerScalarMath() {}

// Destructor.
EulerScalarMath::~EulerScalarMath() {}

MObject EulerScalarMath::inOperation;
MObject EulerScalarMath::inEuler;
MObject EulerScalarMath::inEulerRotOrder;
MObject EulerScalarMath::inScalar;
MObject EulerScalarMath::inResRotOrder;
MObject EulerScalarMath::outEuler;


void* EulerScalarMath::creator(){
    // Maya creator function.
    return new EulerScalarMath();
}

MStatus EulerScalarMath::initialize(){
    /*
    Defines the set of attributes for this node. The attributes declared in this function are assigned
    as static members to TestNode class. Instances of TestNode will use these attributes to create plugs
    for use in the compute() method.
    */
    MStatus status;
    MFnEnumAttribute eAttr;
    MFnUnitAttribute uAttr;
    MFnNumericAttribute nAttr;

    inOperation = eAttr.create("operation", "operation", 0, &status);
    eAttr.addField("No Operation", 0);
    eAttr.addField("Add", 1);
    eAttr.addField("Subtract", 2);
    eAttr.addField("Multiply", 3);
    INPUT_ATTR(eAttr);

    MObject eulerX = uAttr.create("eulerX", "ex", MFnUnitAttribute::kAngle, 0.0, &status);
    MObject eulerY = uAttr.create("eulerY", "ey", MFnUnitAttribute::kAngle, 0.0, &status);
    MObject eulerZ = uAttr.create("eulerZ", "ez", MFnUnitAttribute::kAngle, 0.0, &status);
    inEuler = nAttr.create("euler", "e", eulerX, eulerY, eulerZ, &status);
    INPUT_ATTR(nAttr);

    inEulerRotOrder = eAttr.create("rotateOrderEuler", "roe", 0, &status);
    eAttr.addField("xyz", 0);
    eAttr.addField("yzx", 1);
    eAttr.addField("zxy", 2);
    eAttr.addField("xzy", 3);
    eAttr.addField("yxz", 4);
    eAttr.addField("zyx", 5);
    INPUT_ATTR(eAttr);

    inScalar = nAttr.create("scalar", "scalar", MFnNumericData::kDouble, 0.0, &status);
    INPUT_ATTR(nAttr);

    inResRotOrder = eAttr.create("rotateOrderOutEuler", "rooe", 0, &status);
    eAttr.addField("xyz", 0);
    eAttr.addField("yzx", 1);
    eAttr.addField("zxy", 2);
    eAttr.addField("xzy", 3);
    eAttr.addField("yxz", 4);
    eAttr.addField("zyx", 5);
    INPUT_ATTR(eAttr);

    MObject outEulerX = uAttr.create("outEulerX", "oex", MFnUnitAttribute::kAngle, 0.0, &status);
    MObject outEulerY = uAttr.create("outEulerY", "oey", MFnUnitAttribute::kAngle, 0.0, &status);
    MObject outEulerZ = uAttr.create("outEulerZ", "oez", MFnUnitAttribute::kAngle, 0.0, &status);
    outEuler = nAttr.create("outEuler", "oe", outEulerX, outEulerY, outEulerZ, &status);
    OUTPUT_ATTR(nAttr);

    addAttribute(inOperation);
    addAttribute(inEuler);
    addAttribute(inEulerRotOrder);
    addAttribute(inScalar);
    addAttribute(inResRotOrder);
    addAttribute(outEuler);
    attributeAffects(inOperation, outEuler);
    attributeAffects(inEuler, outEuler);
    attributeAffects(inEulerRotOrder, outEuler);
    attributeAffects(inScalar, outEuler);
    attributeAffects(inResRotOrder, outEuler);

    return status;
}

MStatus EulerScalarMath::compute(const MPlug& plug, MDataBlock& dataBlock){
    /*
    Node computation method:
        * plug is a connection point related to one of our node attributes (either an input or an output).
        * dataBlock contains the data on which we will base our computations.
    */
    if (plug != outEuler)
        return MStatus::kUnknownParameter;
    short operation = dataBlock.inputValue(inOperation).asShort();
    MVector vEuler = dataBlock.inputValue(inEuler).asVector();
    double scalar = dataBlock.inputValue(inScalar).asDouble();
    short eulerRotOrder = dataBlock.inputValue(inEulerRotOrder).asShort();
    short outRotOrder = dataBlock.inputValue(inResRotOrder).asShort();

    MEulerRotation eEuler = EulerScalarMath::createMEulerRotation(vEuler, eulerRotOrder);

    MDataHandle outEulerHandle = dataBlock.outputValue(outEuler);
    MVector vResult, vScalar;
    MEulerRotation eOutEuler;

    switch (operation)
    {
    case 0:
        EulerScalarMath::reorderMEulerRotation(eEuler, outRotOrder);
        vResult = eEuler.asVector();
        outEulerHandle.setMVector(vResult);
        break;
    case 1:
        EulerScalarMath::reorderMEulerRotation(eEuler, outRotOrder);
        vScalar = MVector(
            MAngle(scalar, MAngle::kDegrees).asRadians(),
            MAngle(scalar, MAngle::kDegrees).asRadians(),
            MAngle(scalar, MAngle::kDegrees).asRadians()
        );
        vResult = eEuler.asVector() + vScalar;
        outEulerHandle.setMVector(vResult);
        break;
    case 2:
        EulerScalarMath::reorderMEulerRotation(eEuler, outRotOrder);
        vScalar = MVector(
            MAngle(scalar, MAngle::kDegrees).asRadians(),
            MAngle(scalar, MAngle::kDegrees).asRadians(),
            MAngle(scalar, MAngle::kDegrees).asRadians()
        );
        vResult = eEuler.asVector() - vScalar;
        outEulerHandle.setMVector(vResult);
        break;
    case 3:
        EulerScalarMath::reorderMEulerRotation(eEuler, outRotOrder);
        eOutEuler = eEuler * scalar;
        vResult = eOutEuler.asVector();
        outEulerHandle.setMVector(vResult);
        break;
    }

    outEulerHandle.setClean();

    return MStatus::kSuccess;
}

MEulerRotation EulerScalarMath::createMEulerRotation(MVector& value, short rotOrder){
    /*
    Create an MEulerRotation instance based on a double3 typed values and short typed
    rotation order.
    */
    MEulerRotation::RotationOrder order = MEulerRotation::kXYZ;
    switch (rotOrder)
    {
    case 0:
        order = MEulerRotation::kXYZ;
        break;
    case 1:
        order = MEulerRotation::kYZX;
        break;
    case 2:
        order = MEulerRotation::kZXY;
        break;
    case 3:
        order = MEulerRotation::kXZY;
        break;
    case 4:
        order = MEulerRotation::kYXZ;
        break;
    case 5:
        order = MEulerRotation::kZYX;
        break;
    }
    MEulerRotation eResult = MEulerRotation(value, order);
    return eResult;
}

void EulerScalarMath::reorderMEulerRotation(MEulerRotation& euler, short rotOrder){
    MEulerRotation::RotationOrder order;
    switch (rotOrder)
    {
    case 0:
        order = MEulerRotation::kXYZ;
        break;
    case 1:
        order = MEulerRotation::kYZX;
        break;
    case 2:
        order = MEulerRotation::kZXY;
        break;
    case 3:
        order = MEulerRotation::kXZY;
        break;
    case 4:
        order = MEulerRotation::kYXZ;
        break;
    case 5:
        order = MEulerRotation::kZYX;
        break;
    }
    euler.reorderIt(order);
}

File: core/plugin/__DEV/Python/n_gfUtilEulerScalarMath.py
# -*- coding: utf-8 -*-
"""
Copyright (c) 2019 Giuliano França

Redistribution:
    Something here.

Maya Node:
    [This is a prototype version of the gfUtilAngularMath node. You should be using the related C++ version.]
    This node is a test node and only performs test operations with one input value.

Requirements:
    Maya 2017 or above.

Todo:
    * Convert angle values to radians.

This code supports Pylint. Rc file in project.
"""
import maya.api._OpenMaya_py2 as om2


def maya_useNewAPI():
    """ Function to Maya recognize the use of the Python API 2.0. """
    # pylint: disable=invalid-name, unnecessary-pass
    pass


def INPUT_ATTR(FNATTR):
    """ Configure a input attribute. """
    # pylint: disable=invalid-name
    FNATTR.writable = True
    FNATTR.readable = True
    FNATTR.storable = True
    FNATTR.keyable = True


def OUTPUT_ATTR(FNATTR):
    """ Configure a output attribute. """
    # pylint: disable=invalid-name
    FNATTR.writable = False
    FNATTR.readable = True
    FNATTR.storable = False
    FNATTR.keyable = False


class EulerScalarMath(om2.MPxNode):
    """ Main class of gfUtilEulerScalarMath node. """

    kNodeName = ""
    kNodeClassify = ""
    kNodeID = ""

    inOperation = om2.MObject()
    inEuler = om2.MObject()
    inEulerRotOrder = om2.MObject()
    inScalar = om2.MObject()
    inResRotOrder = om2.MObject()
    outEuler = om2.MObject()

    def __init__(self):
        """ Constructor. """
        om2.MPxNode.__init__(self)

    @staticmethod
    def creator():
        """ Maya creator function. """
        return EulerScalarMath()

    @staticmethod
    def initialize():
        """
        Defines the set of attributes for this node. The attributes declared in this function are assigned
        as static members to EulerScalarMath class. Instances of EulerScalarMath will use these attributes to create plugs
        for use in the compute() method.
        """
        eAttr = om2.MFnEnumAttribute()
        uAttr = om2.MFnUnitAttribute()
        nAttr = om2.MFnNumericAttribute()

        EulerScalarMath.inOperation = eAttr.create("operation", "operation", 0)
        eAttr.addField("No Operation", 0)
        eAttr.addField("Add", 1)
        eAttr.addField("Subtract", 2)
        eAttr.addField("Multiply", 3)
        INPUT_ATTR(eAttr)

        eulerX = uAttr.create("eulerX", "ex", om2.MFnUnitAttribute.kAngle, 0.0)
        eulerY = uAttr.create("eulerY", "ey", om2.MFnUnitAttribute.kAngle, 0.0)
        eulerZ = uAttr.create("eulerZ", "ez", om2.MFnUnitAttribute.kAngle, 0.0)
        EulerScalarMath.inEuler = nAttr.create("euler", "e", eulerX, eulerY, eulerZ)
        INPUT_ATTR(nAttr)

        EulerScalarMath.inEulerRotOrder = eAttr.create("rotateOrderEuler", "roe", 0)
        eAttr.addField("xyz", 0)
        eAttr.addField("yzx", 1)
        eAttr.addField("zxy", 2)
        eAttr.addField("xzy", 3)
        eAttr.addField("yxz", 4)
        eAttr.addField("zyx", 5)
        INPUT_ATTR(eAttr)

        EulerScalarMath.inScalar = nAttr.create("scalar", "scalar", om2.MFnNumericData.kDouble, 0.0)
        INPUT_ATTR(nAttr)

        EulerScalarMath.inResRotOrder = eAttr.create("rotateOrderOutEuler", "rooe", 0)
        eAttr.addField("xyz", 0)
        eAttr.addField("yzx", 1)
        eAttr.addField("zxy", 2)
        eAttr.addField("xzy", 3)
        eAttr.addField("yxz", 4)
        eAttr.addField("zyx", 5)
        INPUT_ATTR(eAttr)

        outEulerX = uAttr.create("outEulerX", "oex", om2.MFnUnitAttribute.kAngle, 0.0)
        outEulerY = uAttr.create("outEulerY", "oey", om2.MFnUnitAttribute.kAngle, 0.0)
        outEulerZ = uAttr.create("outEulerZ", "oez", om2.MFnUnitAttribute.kAngle, 0.0)
        EulerScalarMath.outEuler = nAttr.create("outEuler", "oe", outEulerX, outEulerY, outEulerZ)
        OUTPUT_ATTR(nAttr)

        EulerScalarMath.addAttribute(EulerScalarMath.inOperation)
        EulerScalarMath.addAttribute(EulerScalarMath.inEuler)
        EulerScalarMath.addAttribute(EulerScalarMath.inEulerRotOrder)
        EulerScalarMath.addAttribute(EulerScalarMath.inScalar)
        EulerScalarMath.addAttribute(EulerScalarMath.inResRotOrder)
        EulerScalarMath.addAttribute(EulerScalarMath.outEuler)
        EulerScalarMath.attributeAffects(EulerScalarMath.inOperation, EulerScalarMath.outEuler)
        EulerScalarMath.attributeAffects(EulerScalarMath.inEuler, EulerScalarMath.outEuler)
        EulerScalarMath.attributeAffects(EulerScalarMath.inEulerRotOrder, EulerScalarMath.outEuler)
        EulerScalarMath.attributeAffects(EulerScalarMath.inScalar, EulerScalarMath.outEuler)
        EulerScalarMath.attributeAffects(EulerScalarMath.inResRotOrder, EulerScalarMath.outEuler)

    def compute(self, plug, dataBlock):
        """
        Node computation method:
            * plug is a connection point related to one of our node attributes (either an input or an output).
            * dataBlock contains the data on which we will base our computations.
        """
        # pylint: disable=no-self-use
        if plug != EulerScalarMath.outEuler:
            return om2.kUnknownParameter

        operation = dataBlock.inputValue(EulerScalarMath.inOperation).asShort()
        vEuler = dataBlock.inputValue(EulerScalarMath.inEuler).asVector()
        scalar = dataBlock.inputValue(EulerScalarMath.inScalar).asDouble()
        eulerRotOder = dataBlock.inputValue(EulerScalarMath.inEulerRotOrder).asShort()
        outRotOrder = dataBlock.inputValue(EulerScalarMath.inResRotOrder).asShort()

        eEuler = om2.MEulerRotation(vEuler, eulerRotOder)

        outEulerHandle = dataBlock.outputValue(EulerScalarMath.outEuler)

        if operation == 0:
            eEuler.reorderIt(outRotOrder)
            vResult = eEuler.asVector()
            outEulerHandle.setMVector(vResult)
        elif operation == 1:
            eEuler.reorderIt(outRotOrder)
            vScalar = om2.MVector(
                om2.MAngle(scalar, om2.MAngle.kDegrees).asRadians(),
                om2.MAngle(scalar, om2.MAngle.kDegrees).asRadians(),
                om2.MAngle(scalar, om2.MAngle.kDegrees).asRadians(),
            )
            vResult = eEuler.asVector() + vScalar
            outEulerHandle.setMVector(vResult)
        elif operation == 2:
            eEuler.reorderIt(outRotOrder)
            vScalar = om2.MVector(
                om2.MAngle(scalar, om2.MAngle.kDegrees).asRadians(),
                om2.MAngle(scalar, om2.MAngle.kDegrees).asRadians(),
                om2.MAngle(scalar, om2.MAngle.kDegrees).asRadians(),
            )
            vResult = eEuler.asVector() - vScalar
            outEulerHandle.setMVector(vResult)
        elif operation == 3:
            eEuler.reorderIt(outRotOrder)
            eOutEuler = eEuler * scalar
            vResult = eOutEuler.asVector()
            outEulerHandle.setMVector(vResult)

        outEulerHandle.setClean()
